builder.ts

```diff
@@ -0,0 +1,223 @@
+    try {
+      return (await transformer(value)) || error;
+    } catch (err: any) {
+      return `${error} "${value}": ${err.message}`;
+    }
+  };
+
+const address = (config: { message: string; default?: string }) =>
+  input({
+    ...config,
+    validate: trial((value) => isAddress(value), "Invalid address"),
+  });
+
+const number = (config: { message: string; min: number; default?: string }) =>
+  input({
+    ...config,
+    validate: trial((value) => {
+      const amount = parseInt(value);
+
+      if (isNaN(amount)) return false;
+      if (amount < config.min) return `Value below minimum: ${config.min}`;
+
+      return true;
+    }, "Invalid value"),
+  });
+
+(async () => {
+  console.log();
+  console.log("      #########################################");
+  console.log("      ##         Transaction Builder         ##");
+  console.log("      #########################################");
+
+  console.log(`\n          --- Safe Transaction Options ---\n`);
+
+  const nbTxs = parseInt(
+    await number({
+      message: "How many transactions do you want to batch?",
+      default: "1",
+      min: 1,
+    })
+  );
+  const gasToken = await address({
+    message: `With which ERC20 token do you want to pay gas? (default: ETH)`,
+    default: "0x0000000000000000000000000000000000000000",
+  });
+
+  let txs: MetaTransaction[] = [];
+  for (const index of new Array(nbTxs).fill(0).map((_, index) => index)) {
+    console.log(`\n          --- Transaction #${(index + 1).toString().padStart(2, "0")} ---\n`);
+
+    const prefix = `  `;
+
+    const txType = (await select({
+      message: prefix + `Transaction type:`,
+      choices: [
+        { name: "Call contract", value: TransactionType.callContract },
+        { name: "Transfer funds (ERC20)", value: TransactionType.transferFunds },
+        { name: "Transfer collectible (ERC721/ERC1155)", value: TransactionType.transferCollectible },
+        { name: "Raw transaction", value: TransactionType.raw },
+      ],
+    })) as TransactionType;
+
+    let to = "0x";
+    let data = "0x";
+    let value = "0";
+
+    switch (txType) {
+      case TransactionType.callContract:
+        to = await address({ message: prefix + `Contract address` });
+
+        const functionSelector = await input({
+          message: prefix + `Function selector`,
+          validate: trial((value) => !!FunctionFragment.from(`function ${value}`), "Invalid function selector"),
+        });
+        const fragment = FunctionFragment.from(`function ${functionSelector}`);
+
+        const values = [];
+        for (const { param, index } of fragment.inputs.map((param, index) => ({ param, index }))) {
+          const value = await input({
+            message: prefix + `   ${param.name || "Parameter #" + index} (${param.type})`,
+            validate: (value) => {
+              const process = spawnSync("cast", ["abi-encode", `function(${param.type})`, value]); // Using cast because it reads human JSON.
```

Comment:
Is it the only way to do that? Is it not possible at all with ethers? It's a bit weird

Reply:
It is possible with ethers, but we're losing the human-JSON approach `cast` enables. For example, passing `["0x..."]` to `cast` is straight-forward. I couldn't find a similar approach using `ethers`. The closest is `JSON.parse` (or event `JSON5.parse`), but it coerces large numbers into numbers (not bigints), losing some digits.
So instead of trying to write a new way of validating & transforming inputs, I decided to rely on `cast`, which was built for it

Reply:
Ok clear!